apps/speed.c

```diff
@@ -2299,9 +2334,18 @@ int speed_main(int argc, char **argv)
                                            key16, iv);
                     EVP_CIPHER_CTX_set_padding(loopargs[k].ctx, 0);
                 }
+                switch (EVP_CIPHER_nid(evp_cipher)) {
+                case NID_aes_128_ccm:
+                case NID_aes_192_ccm:
+                case NID_aes_256_ccm:
+                    loopfunc = EVP_Update_loop_ccm;
```

Comment:
The ARIA cipher also supports CCM mode.  This scheme could be extended like this:

```
#ifndef OPENSSL_NO_ARIA
case NID_aria_128_ccm:
case NID_aria_192_ccm:
case NID_aria_256_ccm:
#endif
```

However, I'd greatly prefer to see a more generic way to determine this (using `EVP_CIPHER_mode` perhaps) that doesn't require a list of possible ciphers to be maintained.

```
switch (EVP_CIPHER_mode(evp_cipher)) {
case EVP_CIPH_CCM_MODE:
    loopfunc = EVP_Update_loop_ccm;
    break;
default:
    loopfunc = EVP_Update_loop_ccm;
}
```

It also looks like the documentation in `doc/man3/EVP_EncryptInit.pod` could use a refresh to mention CTR, GCM, CCM, OCB, WRAP and XTS modes

Reply:
agreed. implemented the EVP_CIPHER_mode-based solution.

addressed the doc issue in a separate PR: #4498